src/Abstract/Value.hs

```diff
@@ -19,53 +23,87 @@ envLookup :: Name -> Environment a -> Maybe a
 envLookup = (. unEnvironment) . Map.lookup
 
 envInsert :: Name -> a -> Environment a -> Environment a
-envInsert = (((Environment .) . (. unEnvironment)) .) . Map.insert
+envInsert name value (Environment m) = Environment (Map.insert name value m)
 
 
-data Value l t a
-  = I a
-  | Closure Name t (Environment (l (Value l t a)))
-  deriving (Foldable, Functor, Traversable)
+data Value l
+  = I Prim
+  | Closure Name (Term Prim) (Environment (Address l (Value l)))
 
 
-instance Address l => Eq2 (Value l) where
-  liftEq2 eqT eqA = go
+class Monad m => MonadEnv a m where
+  askEnv :: m (Environment a)
+  localEnv :: (Environment a -> Environment a) -> m b -> m b
+
+instance Reader (Environment a) :< fs => MonadEnv a (Eff fs) where
+  askEnv = ask
+  localEnv = local
+
+
+class Monad m => MonadValue l v t a m where
+  lambda :: (t a -> m v) -> Name -> Type -> t a -> m v
+  app :: (t a -> m v) -> v -> v -> m v
+
+  prim' :: a -> m v
```

Comment:
This typeclass allows us to abstract over lambdas and applications s.t. we can decide their evaluation strategy, specifically whether to evaluate under binders. As a rule of thumb, we generally won’t want to evaluate under binders for purposes of execution, but for typechecking it’s essential—we have to traverse the body of the lambda eagerly.

Thus, the instance for `Value` produces a closure, while the instance for `Type` produces a function type.